src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,621 @@
+                    return assignment;
+                }
+            }
+            else
+            {
+                refAssignmentSourceIsLocal = null;
+                refAssignmentSourceIndex = null;
+            }
+
+            if (!TryGetLocalOrParameterInfo(original.Left, out var targetSymbol, out var targetType, out var targetIndex))
+            {
+                Debug.Fail("Must be local or parameter");
+            }
+
+            var logger = GetLocalOrParameterStoreLogger(targetType, targetSymbol, refAssignmentSourceIsLocal, original.Syntax);
+            if (logger is null)
+            {
+                return assignment;
+            }
+
+            return _factory.Sequence(new[]
+            {
+                _factory.Call(
+                    receiver: _factory.Local(_scope.ContextVariable),
+                    logger,
+                    MakeStoreLoggerArguments(logger.Parameters[0], targetSymbol, targetType, assignment, refAssignmentSourceIndex, targetIndex))
+            }, VariableRead(targetSymbol));
+        }
+
+        private bool TryGetLocalOrParameterInfo(BoundNode node, [NotNullWhen(true)] out Symbol? symbol, [NotNullWhen(true)] out TypeSymbol? type, [NotNullWhen(true)] out BoundExpression? indexExpression)
+        {
+            if (node is BoundLocal { LocalSymbol: var localSymbol })
+            {
+                symbol = localSymbol;
+                type = localSymbol.Type;
+                indexExpression = _factory.LocalId(localSymbol);
+                return true;
+            }
+
+            if (node is BoundParameter { ParameterSymbol: var parameterSymbol })
+            {
+                Debug.Assert(!parameterSymbol.IsDiscard);
+
+                symbol = parameterSymbol;
+                type = parameterSymbol.Type;
+                indexExpression = _factory.ParameterId(parameterSymbol);
+                return true;
+            }
+
+            symbol = null;
+            indexExpression = null;
+            type = null;
+            return false;
+        }
+
+        private ImmutableArray<BoundExpression> MakeStoreLoggerArguments(
+            ParameterSymbol parameter,
+            Symbol targetSymbol,
+            TypeSymbol targetType,
+            BoundExpression value,
+            BoundExpression? refAssignmentSourceIndex,
+            BoundExpression index)
+        {
+            if (refAssignmentSourceIndex != null)
+            {
+                return ImmutableArray.Create(_factory.Sequence(new[] { value }, refAssignmentSourceIndex), index);
+            }
+
+            Debug.Assert(parameter.RefKind == RefKind.None);
+
+            if (parameter.Type.IsVoidPointer() && !targetType.IsPointerOrFunctionPointer())
+            {
+                // address of assigned value to be passed to LogStore*Unmanaged:
+                Debug.Assert(!parameter.Type.IsManagedTypeNoUseSiteDiagnostics);
+
+                var addressOf = value is BoundLocal or BoundParameter ?
```

Comment:
>value [](http://example.com/codeflow?start=32&length=5)

I'm confused why we're using `value` here (which looks to be the entire assignment in the caller above) rather than `assignment.Left`.

Reply:
`value` could be either an assignment, or direct parameter or local access (e.g. when we are logging the initial value of a parameter, or the value of a variable after a call with `out var`). If we have the variable itself we just take an address, otherwise we need to emit the assignment, read the symbol and take its address.